KAFKA-6981: Move the error handling configuration properties into the ConnectorConfig and SinkConnectorConfig classes

wicknicks opened the pull request:
Move the error handling configuration properties into the ConnectorConfig and SinkConnectorConfig classes, and refactor the tests and classes to use these new properties.

Testing: Unit tests and running the connect-standalone script with a file sink connector.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rhauch committed:
KAFKA-6981: Move the error handling configuration properties into the ConnectorConfig and SinkConnectorConfig

Work in progress

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6981: Refactor classes and tests to use properties from ConnectorConfig or SinkConnectorConfig

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6981: Fix whitespace issues

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
retest this please

wicknicks commented:
@rhauch, @mageshn this PR should contain all the error handling configs in ConnectorConfig or SinkConnectorConfig. please let me know what you think. thanks!

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DeadLetterQueueReporter.java:
```diff
@@ -124,31 +110,18 @@ public void report(ProcessingContext context) {
         ProducerRecord<byte[], byte[]> producerRecord;
         if (originalMessage.timestamp() == RecordBatch.NO_TIMESTAMP) {
-            producerRecord = new ProducerRecord<>(config.topic(), null,
+            producerRecord = new ProducerRecord<>(connConfig.dlqTopicName(), null,
```
Normally we would not want to read something from the config for every record that we process, and would instead want to read the config once and store the topic name in a final member to use many times. 

But in this case it might be okay since this is only handling errors, which should not be as frequent. 

Thoughts?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java:
```diff
@@ -61,24 +55,12 @@
     // for testing only
-    public static final RetryWithToleranceOperator NOOP_OPERATOR = new RetryWithToleranceOperator();
+    public static final RetryWithToleranceOperator NOOP_OPERATOR = new RetryWithToleranceOperator(
+            ConnectorConfig.ERRORS_RETRY_TIMEOUT_DEFAULT, ConnectorConfig.ERRORS_RETRY_MAX_DELAY_DEFAULT, ToleranceType.NONE, Time.SYSTEM);
```
Should this move to the test code? Where is this used in the non-test code?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperator.java:
```diff
@@ -90,27 +72,31 @@
-    public RetryWithToleranceOperator(Time time) {
-        this.time = time;
+    public RetryWithToleranceOperator(ConnectorConfig connConfig, Time time) {
+        this(connConfig.errorRetryTimeout(), connConfig.errorMaxDelayInMillis(), connConfig.errorToleranceType(), time);
```
Could we keep just one of these constructors? We could keep the constructor that takes pass a config and have the tests pass a mock with mocked methods for these getters, or we keep the last form and change where this constructor is called. 

I think it would prefer the latter. We have a number of classes that are similar, so there is lots of prior art.

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java:
```diff
@@ -170,16 +170,16 @@ public void testConfigValidationMissingName() {
         assertEquals(2, result.errorCount());
         // Base connector config has 8 fields, connector's configs add 2
-        assertEquals(10, result.values().size());
+        assertEquals(15, result.values().size());
```
The comment above this line is no longer correct

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceOperatorTest.java:
```diff
@@ -243,56 +241,54 @@ public void testBackoffLimit() {
-        assertEquals(configuration.retryTimeout(), 100);
+    public void testSetConfigs() {
+        ConnectorConfig configuration;
+        configuration = config(singletonMap("errors.retry.timeout", "100"));
```
Are we purposefully avoiding the use of the config constants, for example to check for changes in the constants?

rhauch commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/errors/ErrorReporterTest.java:
```diff
@@ -184,9 +187,20 @@ private ProcessingContext processingContext() {
+        props.put(ConnectorConfig.NAME_CONFIG, "test");
+        props.put(ConnectorConfig.CONNECTOR_CLASS_CONFIG, SinkTask.class.getName());
+        props.putAll(configProps);
+        return new ConnectorConfig(plugins, props);
```
Since SinkConnectorConfig extends ConnectorConfig, could we just use the first helper method everywhere?

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -106,6 +108,40 @@
+    public static final String ERRORS_RETRY_MAX_DELAY_DOC = "The maximum duration in milliseconds between consecutive retry attempts. " +
+            "Jitter will be added to the delay once this limit is reached to prevent thundering herd issues.";
+
+    public static final String ERRORS_TOLERANCE_CONFIG = "errors.allowed.max";
```
I think this config property key seems a misfit, and probably reflects an earlier incantation of the design before KIP acceptance. It might be worth - in a separate PR - renaming this to something like `errors.tolerance` to better align with its purpose.

wicknicks replied:
You are right. I think it should be ok to update the current KIP. Would be better to have the first release contain the most reasonable name. Thoughts?

rhauch replied:
No, we have to fix it before AK 2.0. Once it is in a released API, we can’t change it.

wicknicks replied:
Sounds good. I'll update this in another PR after this is merged. Thanks!

rhauch replied:
We'll need a separate AK issue, then.

kkonstantine replied:
The PR title sounds like something that is reasonable to contain a fix on configs if you need an opinion.

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java:
```diff
@@ -42,9 +43,19 @@
         .define(TOPICS_CONFIG, ConfigDef.Type.LIST, TOPICS_DEFAULT, ConfigDef.Importance.HIGH, TOPICS_DOC, COMMON_GROUP, 4, ConfigDef.Width.LONG, TOPICS_DISPLAY)
-        .define(TOPICS_REGEX_CONFIG, ConfigDef.Type.STRING, TOPICS_REGEX_DEFAULT, new RegexValidator(), ConfigDef.Importance.HIGH, TOPICS_REGEX_DOC, COMMON_GROUP, 4, ConfigDef.Width.LONG, TOPICS_REGEX_DISPLAY);
+        .define(TOPICS_REGEX_CONFIG, ConfigDef.Type.STRING, TOPICS_REGEX_DEFAULT, new RegexValidator(), ConfigDef.Importance.HIGH, TOPICS_REGEX_DOC, COMMON_GROUP, 4, ConfigDef.Width.LONG, TOPICS_REGEX_DISPLAY)
+        .define(DLQ_TOPIC_NAME_CONFIG, ConfigDef.Type.STRING, DLQ_TOPIC_DEFAULT, Importance.MEDIUM, DLQ_TOPIC_NAME_DOC, ERROR_GROUP, 6, ConfigDef.Width.MEDIUM, DLQ_TOPIC_DISPLAY);
```
Should this be added to the Errors group?

wicknicks replied:
I am ok to put the property in this group. The only worrisome point is the hardcoded `6` here. Any way to get around this?

rhauch replied:
I don’t know of an easy way to do that.

rhauch replied:
Well, one way is to iterate through the ConfigKeys and count the number of existing keys in the group. But there's no existing utility or method to do this.

wicknicks replied:
Can we make orderInErrorGroup a protected static field in `ConnectorConfig` and use `++orderInErrorGroup` instead of `6`? Same for `orderInGroup`?

rhauch replied:
That seems even more brittle, doesn't it. :-)

wicknicks replied:
Ok. I'll leave it as it is then. Let me know if you have any further comments.

rhauch replied:
Yeah, ConfigDef doesn't break if there are duplicate `orderInGroup` values, and it tries to [order them in a somewhat orderly fashion](https://github.com/apache/kafka/blob/38e9958d6e5d2224fbb9a54d04b1505ad561e8a3/clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java#L1246-L1270). This is why I think this is less brittle than trying to manage those orders across multiple Configuration classes.

wicknicks replied:
Yeah, makes sense. That's how the `TOPICS_CONFIG` and `TOPICS_REGEX_CONFIG` work without any issues. Thanks, Randall.

rhauch reviewed (commented):
Looks good! A few comments and suggestions below.

wicknicks committed:
KAFKA-6981: Address reviewer comments

Jira: KAFKA-6981
Signed-off-by: Arjun Satish <arjun@confluent.io>

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DeadLetterQueueReporter.java:
```diff
@@ -49,22 +47,14 @@
+        String topic = connConfig.dlqTopicName();
 
         try (AdminClient admin = AdminClient.create(workerConfig.originals())) {
             if (!admin.listTopics().names().get().contains(topic)) {
```
BTW, the following line seems problematic for any Kafka cluster with fewer than 3 brokers:

    NewTopic schemaTopicRequest = new NewTopic(topic, DLQ_NUM_DESIRED_PARTITIONS, DLQ_MAX_DESIRED_REPLICATION_FACTOR);

We should likely allow this to be set via the config, or at a minimum we should log a much better error message and suggest users create the topic on their own. The latter is more error prone and users are likely to not know this up front, so doing that _and_ offering the configuration approach is the better option. Probably should be done in a separate AK issue.

wicknicks replied:
In the first iteration of this code, we checked the number of brokers available on the cluster before setting the replication factor. But that has apparently caused issues in the past where having fewer brokers during the time this code was called, created a topic with less than desired replication factor. Hence, we decided to simply put this number as it is.

A config would mean changes to the KIP? IIRC, we had it in the first iteration of the KIP, but reviewers wanted a simpler first version.

rhauch replied:
I think there are three options:
1. Leave as is, but improve the error message and suggest users manually create the topic. This has usability implications, though I don't expect too many demos will enable DLQ. This does not require changing the KIP.
2. Add one configuration to set the replication factor. Like the other Connect settings that deal with replication factors, this would be low priority and should default to 3, which should be fine for many production use cases when DLQ is implemented. The error message should still be modified to be useful, but that error should happen less frequently. This means updating the KIP.
3. Never even attempt to create the topic, so fail if it does not exist and improve the error message with more useful detail. This means updating the KIP to at least suggest that the topic will not be created.

IMO, option 2 is the most usable, but it still requires users with small clusters (e.g., a dev environment) that are enabling DLQ to set this. Options 1 and 3 are both less useable.

We should probably create a Jira issue and discuss it there, but it needs to be done quickly and it should be marked a blocker.

kkonstantine replied:
I had pointed out in the initial review that the heuristic to guess the desired RF seemed of little value. I'm also for reconsidering.

wicknicks replied:
@kkonstantine there is no heuristic here, do you mean have the replication factor come in from a config field?

rhauch reviewed (commented):
One more pass...

rhauch commented:
@wicknicks the build errors are unused imports.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java:
```diff
@@ -42,9 +43,19 @@
+        "result in an error when processed by this sink connector, transformations, or converter. The topic name is blank by default, " +
+        "which means that no messages are to be recorded in the DLQ.";
+    public static final String DLQ_TOPIC_DEFAULT = "";
+    private static final String DLQ_TOPIC_DISPLAY = "DLQ Topic Name";
```
Let's avoid `DLQ` in the display name too. `DeadLetterQueue Topic Name` should be fine.

wicknicks committed:
KAFKA-6981: Fix build errors

Signed-off-by: Arjun Satish <arjun@confluent.io>

kkonstantine reviewed (commented):
Two small comments for now by me.

wicknicks committed:
KAFKA-6981: Expand DLQ acronym

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@kkonstantine: I've addressed your comments. If you are ok with what we have now, let's go ahead merge this PR. So we can address the other issues in future PRs.

kkonstantine reviewed (approved):
I'm 👍 w.r.t my brief comments.

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -106,6 +108,40 @@
+    public static final String ERRORS_TOLERANCE_ALL = "all";
+    public static final String ERRORS_TOLERANCE_DEFAULT = ERRORS_TOLERANCE_NONE;
+    public static final String ERRORS_TOLERANCE_DOC = "Behavior for tolerating errors during connector operation. NONE is the default " +
+            "and signals that any error will result in a connector task failure, and this matches the old behavior. ALL changes the " +
```
nit: matches the old behavior is very relative

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -106,6 +108,40 @@
+    public static final String ERRORS_LOG_ENABLE_CONFIG = "errors.log.enable";
+    public static final String ERRORS_LOG_ENABLE_DISPLAY = "Log Errors";
+    public static final boolean ERRORS_LOG_ENABLE_DEFAULT = false;
+    public static final String ERRORS_LOG_ENABLE_DOC = "Whether to output in the Connect application log each error and the context " +
```
The first part of the doc sounds incomplete.

mageshn commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -505,7 +505,8 @@ private WorkerTask buildWorkerTask(ClusterConfigState configState,
         } else if (task instanceof SinkTask) {
             TransformationChain<SinkRecord> transformationChain = new TransformationChain<>(connConfig.<SinkRecord>transformations(), retryWithToleranceOperator);
-            retryWithToleranceOperator.reporters(sinkTaskReporters(id, connConfig, errorHandlingMetrics));
+            SinkConnectorConfig sinkConfig = new SinkConnectorConfig(plugins, connConfig.originalsStrings());
```
connConfig should already be an instance of SinkConnectorConfig in this condition. Wondering if this needs to be reconstructed here again.

wicknicks replied:
Thanks, @mageshn! we need the dlq topic name from the sink connector config. Hence, we definitely have to create this object here? re-using the `connConfig` would miss out on those properties. Hope I'm not missing anything.

rhauch replied:
Right. The method that calls this instantiates `ConnectorConfig`, not `SinkConnectorConfig`, so it needs to be constructed here.

mageshn replied:
@rhauch That's right. I missed that. I assumed, the right instance would be constructed in the first place :)

mageshn reviewed (commented):
This is great @wicknicks. Had few minor comments.

wicknicks committed:
KAFKA-6981: Reword documentation for configs

Signed-off-by: Arjun Satish <arjun@confluent.io>

rhauch commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -125,15 +125,14 @@
-            "and signals that any error will result in a connector task failure, and this matches the old behavior. ALL changes the " +
-            "behavior to skip over problem records.";
+    public static final String ERRORS_TOLERANCE_DOC = "Behavior for tolerating errors during connector operation. 'NONE' is the default value " +
+            "and signals that any error will result in an immediate connector task failure. 'ALL' changes the behavior to skip over problematic records.";
```
Aren't the actual values that are used in the recommender are lowercase? If so, we should be consistent here.

wicknicks replied:
Either work (we ignore the case). but I'll update the code though.

mageshn replied:
In that case, we can just reuse the enum instead of defining another string constant?

wicknicks committed:
KAFKA-6981: Lowercase values for ERROR_TOLERANCE_DOC

Signed-off-by: Arjun Satish <arjun@confluent.io>

rayokota commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -106,6 +108,39 @@
+    public static final String ERRORS_RETRY_TIMEOUT_CONFIG = "errors.retry.timeout";
+    public static final String ERRORS_RETRY_TIMEOUT_DISPLAY = "Retry Timeout for Errors";
+    public static final int ERRORS_RETRY_TIMEOUT_DEFAULT = 0;
+    public static final String ERRORS_RETRY_TIMEOUT_DOC = "The maximum duration in milliseconds for that a failed operation " +
```
Extra "for" in "milliseconds for that"

rayokota commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java:
```diff
@@ -106,6 +108,39 @@
+    public static final String ERRORS_TOLERANCE_ALL = "all";
+    public static final String ERRORS_TOLERANCE_DEFAULT = ERRORS_TOLERANCE_NONE;
+    public static final String ERRORS_TOLERANCE_DOC = "Behavior for tolerating errors during connector operation. 'none' is the default value " +
+            "and signals that any error will result in an immediate connector task failure. 'all' changes the behavior to skip over problematic records.";
```
Maybe use a semicolon instead:  "task failure; 'all' changes..."

rayokota commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java:
```diff
@@ -42,9 +43,19 @@
+
+    public static final String DLQ_TOPIC_NAME_CONFIG = DLQ_PREFIX + "topic.name";
+    public static final String DLQ_TOPIC_NAME_DOC = "The name of the topic to be used as the dead letter queue (DLQ) for messages that " +
+        "result in an error when processed by this sink connector, transformations, or converter. The topic name is blank by default, " +
```
Maybe "when processed by this sink connector, or any its transformations or converters."

rayokota reviewed (commented):
Looks great @wicknicks !  Just a few suggested changes to documentation

wicknicks committed:
KAFKA-6981: Use the ToleranceType enum to define defaults

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6981: Reword documentation

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks committed:
KAFKA-6981: Introduce value method in ToleranceType enum

Signed-off-by: Arjun Satish <arjun@confluent.io>

wicknicks commented:
@mageshn and @rayokota I think I've addressed all your comments. Could you please take another pass at the latest commits? Thank you!

rayokota reviewed (approved):
LGTM

mageshn reviewed (approved):
Thanks, @wicknicks. LGTM.

rhauch reviewed (approved):
LGTM.

wicknicks commented:
@ewencp Can you take a look at this PR? If it looks good, could we please merge it?

ewencp reviewed (approved):
LGTM, merging to trunk for 2.0.0
